consensus/src/state_computer.rs

```diff
@@ -86,10 +88,19 @@ impl ExecutionProxy {
             async_state_sync_notifier: tx,
             async_commit_notifier: commit_tx,
             validators: Mutex::new(vec![]),
+            data_manager,
+        }
+    }
+
+    async fn get_transactions_from_payload(&self, payload: Payload) -> Vec<SignedTransaction> {
+        match payload {
+            Payload::DirectMempool(txns) => txns,
+            Payload::InQuorumStore(poss) => self.data_manager.get_data(poss).await,
```

Comment:
state_computer shouldn't have to understand the implementation of Payload.

This match can be moved directly into data_manager.get_data(Payload) or alternatively we could move the get logic to payload.get_data(DataManager).

Somewhat related, we need to clarify if payload.is_empty() and payload.len() need to count the batches or txns? Is there any concept of an empty batch, or would a batch never be created if it's empty?

Reply:
Good point.